code/iaas/ha/src/main/java/io/cattle/platform/ha/monitor/impl/PingInstancesMonitorImpl.java

```diff
@@ -56,12 +90,18 @@ public void pingReply(Ping ping) {
         }
 
         long agentId = Long.parseLong(ping.getResourceId());
-        Map<String, String> knownInstances = instanceCache.getUnchecked(agentId);
+        Map<String, Instance> knownInstances = instanceCache.getUnchecked(agentId);
+
+        Agent agent = objectManager.loadResource(Agent.class, agentId);
+        Map<String, Host> hosts = agentDao.getHosts(agent.getId());
+        Host host = hosts.get(reportedInstances.hostUuid);
 
-        if (handleUnreportedKnownInstances(agentId, knownInstances, reportedInstances, true)) {
+        try {
+            syncContainers(knownInstances, reportedInstances, agent.getAccountId(), agent.getId(), host.getId(), true);
+        } catch (ContainersOutOfSync e) {
             knownInstances = load(agentId);
             instanceCache.put(agentId, knownInstances);
-            handleUnreportedKnownInstances(agentId, knownInstances, reportedInstances, false);
+            syncContainers(knownInstances, reportedInstances, agent.getAccountId(), agent.getId(), host.getId(), false);
```

Comment:
Good question. Yes, theoretically, the syncActions can be out of date by the time they're executed. We do our best to avoid this, but there is a small chance of it happening still.

To mitigate, we:
 - don't do any syncing for an instance if it is in a transitioning state
 - we do additional state checks in ContainerEventCreate as well.
- we allow processes scheduled via ContainerEventCreate to be rejected if the container turns out to be in a different state (by catching and swallowing the ProcessCancelException which occurs when you schedule a process and the resource is not in a valid start state for that process)
 - We do as much as possible via ContainerEventCreate since those are no-op events that won't affect the actual container state. The exception to this is system containers, which need ha restarts, but rancher is the system of record for them, so what we decide to do in the check should be accurate.

So, even with all that, there is small chance we could miss schedule something. In this case, Darren said that his philosophy is that that's ok because it's a small chance and the containers should eventually sync back up (possibly having to wait up to 15 minutes for the knownInstances cache to be invalidated).